### DISCUSSION TITLE
One possible solution for postgres expression comparsion

### DISCUSSION THREAD

**Discussion by JabberWocky-22** [sqlalchemy/alembic#1488] [Ideas]
# Issue
In postgres, people may create an expression index or check constraint. However the expression used in database is not exactly same as user's input. This doesn't affect actual usage at all, but has significant impact on expresion comparsion. There are always auto type cast and sometime even some transformation. This makes index and check constraint(may be in the future) comparasion 
almost impossible.

releated issues: https://github.com/sqlalchemy/alembic/issues/1390, https://github.com/sqlalchemy/alembic/issues/1462

For example, if I have a table as follows:
```sql
CREATE TABLE test (id INT, TYPE varchar, extra jsonb);
CREATE INDEX ix_test_info on test ((extra->'user'->>'info'));
ALTER TABLE test ADD CONSTRAINT ck_test_type CHECK (type in ('a', 'b', 'c'));
```
And take a look at the expressions:
index: `(extra -> 'user'::text) ->> 'info'::text`
constraint: `type::text = ANY (ARRAY['a'::character varying, 'b'::character varying, 'c'::character varying]::text[])`

For index, some regex to remove unwanted type cast and space migth work. As for constraint? I want to scream!
Well, It's definitely impossible to compare expression in python(hope I'm wrong). So is it possible to solve it in database?

# Solution

Actually creating one in database and read from result did work, but requires permission for ddl(maybe fine for autogenerate?)

Another possible way is to use `explain`
```sql
EXPLAIN
SELECT * FROM test WHERE (extra->'user'->>'info') IS NOT NULL;
```
The result is:
```
Seq Scan on test  (cost=0.00..22.75 rows=846 width=68)
  Filter: (((extra -> 'user'::text) ->> 'info'::text) IS NOT NULL)
```
Then filter unwanted part, it becomes: `(extra -> 'user'::text) ->> 'info'::text`
It's exactly the same!
And for constraint:
```sql
EXPLAIN
SELECT * FROM test WHERE (type IN ('a','b','c')) IS NULL;
```
And result:
```
Seq Scan on test  (cost=0.00..21.69 rows=4 width=68)
  Filter: (((type)::text = ANY ('{a,b,c}'::text[])) IS NULL)
```
Well, this one doesn't wrok. But if we use this expression to create constraint instead of the original one:
```sql
ALTER TABLE test ADD CONSTRAINT ck_test_type2 CHECK ((type)::text = ANY ('{a,b,c}'::text[]))
```
And the expression in database becomes: `type::text = ANY ('{a,b,c}'::text[])`
Everyhing looks fine!

# Summary
Use explain to do the cast in postgres, so the expression could be same to compare. This shold be done before create statment and comparsion.

Have a nice day.

**Comment by JabberWocky-22** [comment]
It might also contains dll operation in a single migration, so the explain migth fail due to column missing.
It doesn't matter for comparsion, since the column doesn't exist, the index can not be same!
However, missing column will block the explain statement. I have no idea how to resolve this.

Also, this might not work between different versions of postgres.

**Comment by CaselIT** [comment]
Hi,

Thanks an interesting idea, I had not though of using explain.
Sadly as far as I know there is no way of parting an expression into a [node_tree](https://pgpedia.info/p/pg_node_tree.html) that's how PG represents the expressions of the indexes and check constraints

Another alternative is to create a dummy table to create a dummy index against, since there does not seem to be a way of defining an index as "not valid", so that can't be created against a live table (since that could take a lot of time). An alternative to a dummy table is a an empty materialized view `select * from source_table where 1=0` or using the `with no data` clause)
Both of these options are not great though since they required create grant and may leave leftover types around, but I don't see much other options without writing/using an expression parser.

@zzzeek did you ever think of other alternatives?

**Reply by zzzeek** [reply]
I think we just need to write a more comprehensive parser for PostgreSQL's constraint information.   it's not that hard, just a little more work than usual

**Reply by CaselIT** [reply]
seems that the create temp table is granted by default, so that may remove the main issue of using the tables instead. 
Ref: https://www.postgresql.org/docs/current/ddl-priv.html
> For other types of objects, the default privileges granted to PUBLIC are as follows: CONNECT and TEMPORARY (create temporary tables) privileges for databases;

@zzzeek regarding a parser the main issue is that PG is really strange in what it can do.
see for example `type in ('a', 'b', 'c')` -> `type::text = ANY (ARRAY['a'::character varying, 'b'::character varying, 'c'::character varying]::text[])` but may also be `type::text = ANY ('{a,b,c}'::text[])`

**Reply by CaselIT** [reply]
btw I'm currently looking at would a parser look like here
